src/mlpack/core/optimizers/gradient_descent/gradient_descent_impl.hpp

```diff
@@ -0,0 +1,80 @@
+/**
+ * @file gradient_descent_impl.hpp
+ * @author Sumedh Ghaisas
+ *
+ * Simple gradient descent implementation.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_GRADIENT_DESCENT_GRADIENT_DESCENT_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_GRADIENT_DESCENT_GRADIENT_DESCENT_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "gradient_descent.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename FunctionType>
+GradientDescent<FunctionType>::GradientDescent(
+    FunctionType& function,
+    const double stepSize,
+    const size_t maxIterations,
+    const double tolerance) :
+    function(function),
+    stepSize(stepSize),
+    maxIterations(maxIterations),
+    tolerance(tolerance)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename FunctionType>
+double GradientDescent<FunctionType>::Optimize(
+    arma::mat& iterate)
+{
+  // To keep track of where we are and how things are going.
+  double overallObjective = function.Evaluate(iterate);
+  double lastObjective = DBL_MAX;
+
+  // Now iterate!
+  arma::vec gradient(iterate.n_cols);
```

Comment:
This should be `arma::mat` not `arma::vec`, and you should use both the number of rows and columns to initialize it.  Otherwise it will fail if we use an objective function that uses a 2-d matrix instead of a 1-d vector.